File: xjakt00-adventura/PrikazVolej.java
/*******************************************************************************
 * Příkaz 'volej' dovoluje hráči zvát ostatní osoby na party.
 *
 * @author    Tomáš Jakubčin (xjakt00)
 * @version   pro školní rok 2010/11
 */
public class PrikazVolej implements IPrikaz
{
    private static final String NAZEV = "volej";
    private HerniPlan plan;
    private Batoh batoh;
    private Prostor pryc;
    private Hra hra;

    /***************************************************************************
     *  Konstruktor třídy. Kromě plánu a batohu se zde také vytváří také prostor
     *  a v něm i několik osob - kamarádů.
     *  
     *  @param batoh odkaz na hráčův batoh.
     *  @param plan  odkaz na herní plán.
     *  @param hra   odkaz na hru.
     */ 
    public PrikazVolej(Batoh batoh, HerniPlan plan, Hra hra) {
        this.plan = plan;
        this.batoh = batoh;
        this.hra = hra;
        pryc = new Prostor ("pryc", "Prostor urceny pro kamarady z telefonu apod.", false);
        pryc.setOsoba (new Osoba ("Michal", "kamarad", null, null, null, 0, false));
        pryc.setOsoba (new Osoba ("Roman", "kamarad", null, null, null, 0, false));
        pryc.setOsoba (new Osoba ("Tomas", "kamarad", null, null, null, 0, false));
        pryc.setOsoba (new Osoba ("Lubor", "kamarad", null, null, null, 0, false));
        pryc.setOsoba (new Osoba ("Petr", "kamarad", null, null, null, 0, false));
        pryc.setOsoba (new Osoba ("Erik", "kamarad", null, null, null, 0, false));
        pryc.setOsoba (new Osoba ("Martin", "kamarad", null, null, null, 0, false));
        pryc.setOsoba (new Osoba ("Bohous", "kamarad", null, null, null, 0, false));
        pryc.setOsoba (new Osoba ("Josef", "kamarad", null, null, null, 0, false));
        pryc.setOsoba (new Osoba ("Karel", "kamarad", null, null, null, 0, false));
    }

    /**
     *  Metoda proveď musí zkontrolovat parametry a zjistit, jestli je možné
     *  se danému člověku dovolat. Poté změní jeho proměnnou 'pozvana' na true.
     */ 
    public String proved (String... parametry) {
        if (parametry.length == 0) {
            return "Musis zadat komu volat. Pokud ani sam nevis, komu muzes zavolat, napis 'volej koho'.";
        }
        String jmenoSKymMluvit = parametry [0];
        if (jmenoSKymMluvit.equals("koho")) {
            return pryc.seznamVeciAOsob();
        }
        else {
            Prostor kdeJsme = plan.getAktualniProstor();
            Osoba osoba = pryc.vyberOsobu(jmenoSKymMluvit);
            if (osoba == null)
                return "Teto osobe nemuzes volat. Pokud ani sam nevis, komu muzes zavolat, napis 'volej koho'.";
            if (osoba.getPozvana() == true)
                return "Tato osoba je jiz pozvana.";
            if (batoh.vyberVec("mobil") == null) {
                if (!(kdeJsme.getNazev().equals("byt"))) {
                    return "Abys mohl volat, musis byt bud doma, anebo mit mobil.";
                }
            }
            if (plan.getCas()-10 > 0) {
                plan.setCas(10);
                osoba.setPozvana(true);
                plan.setLidiPozvanych(1);
                return "Osoba "+osoba.getJmeno()+" byla pozvana na party.";
            }
            else {
                plan.konecHry();
                return "";
            }
        }
    }

    /**
     *  Metoda vrací název příkazu (slovo které používá hráč pro jeho vyvolání)
     *  
     *  @ return nazev prikazu
     */
    public String getNazev() {
        return NAZEV;
    }
}

File: src/Adventura/PrikazVolej.cpp
#include "PrikazVolej.h"

#include "Hra.h"

PrikazVolej::PrikazVolej(const std::string & nazev, Hra * hra, HerniPlan * herni_plan, Batoh * batoh)
	: IPrikaz(nazev), odkaz_na_hru(hra), odkaz_na_plan(herni_plan), odkaz_na_batoh(batoh),
	telefonni_seznam("seznam", "seznam", false)
{
	telefonni_seznam.pridej_osoby(Osoba("michal", "kamarad"));
	telefonni_seznam.pridej_osoby(Osoba("roman", "kamarad"));
	telefonni_seznam.pridej_osoby(Osoba("tomas", "kamarad"));
	telefonni_seznam.pridej_osoby(Osoba("lubor", "kamarad"));
	telefonni_seznam.pridej_osoby(Osoba("petr", "kamarad"));
	telefonni_seznam.pridej_osoby(Osoba("erik", "kamarad"));
	telefonni_seznam.pridej_osoby(Osoba("bohous", "kamarad"));
	telefonni_seznam.pridej_osoby(Osoba("josef", "kamarad"));
	telefonni_seznam.pridej_osoby(Osoba("karel", "kamarad"));
}

PrikazVolej::~PrikazVolej()
{
}

void PrikazVolej::proved(const std::vector<std::string>& parametry)
{
	if (parametry.empty())
	{
		std::cout << "Musis zadat komu volat. Pokud ani sam nevis, komu muzes zavolat, napis 'volej koho'.\n";
	}
	else if (parametry.size() > 1)
	{
		std::cout << "Nelze volat vice lidem najednou.\n";
	}
	else
	{
		if (parametry[0] == "koho")
		{
			telefonni_seznam.vypis_osoby();
		}
		else
		{
			auto aktualni_prostor = odkaz_na_plan->get_aktualni_prostor();
			bool mam_mobil = odkaz_na_batoh->mam_vec("mobil");

			// bud jsem doma nebo mam mobil
			if (aktualni_prostor->get_jmeno() == "byt" or mam_mobil == true)
			{
				bool mam_v_seznamu = telefonni_seznam.je_osoba_v_prostoru(parametry[0]);
				if (mam_v_seznamu == true)
				{
					bool result = odkaz_na_plan->pozvi_cloveka(Osoba(parametry[0], "kamarad"));
					if (result == true)
					{
						std::cout << "Osoba " << parametry[0] << " byla pozvana na party.\n";
					}
					else
					{
						std::cout << "Tato osoba je jiz pozvana.\n";
					}
				}
				else
				{
					std::cout << "Teto osobe nemuzes volat, protoze neni v tvem telefonnim seznamu."
						<< "Pokud ani sam nevis, komu muzes zavolat, napis 'volej koho'\n";
				}
			}
			else
			{
				std::cout << "Abys mohl volat, musis byt bud doma nebo mit mobil.\n";
			}
		}
	}
}

std::string PrikazVolej::nazev(void) const
{
	return IPrikaz::jmeno_prikazu;
}